Fixed #32539 -- Added toggleable facet filters to ModelAdmin.

sarahboyce opened the pull request:
Ticket: https://code.djangoproject.com/ticket/32539

Facet counts can be added to all filters on a ModelAdmin by either setting ``show_facets`` to ``True`` or by adding the query parameter `?...&_facets=1`

emarhin reviewed (approved):
👍

felixxm commented:
@simonw What do you think?

carltongibson reviewed (commented):
Hey @sarahboyce — thanks for this. It's looks really quite good. 

I uploaded a test project (with a DB, you can just `runserver` on) here: https://github.com/carltongibson/django_ticket_32539

It has a couple of the requisite filters, and sure enough: 

![Screenshot 2023-02-02 at 10 02 28](https://user-images.githubusercontent.com/64686/216278879-c794ea03-626d-40be-84d6-4b9c5b24c258.png)

So 💃. 

I think the docs don't really pull-out what's on offer here, or exactly what's going on when defining the filter. (First coffee and all but, I knew what I was hoping for this morning and was still a bit 🤔 as I was putting it together.) I wonder if we can make it clearer. 

* Can we make this easier still? (The lookup implementation feels a bit clunky)
* Is faceting going to get expensive (too suddenly — is there a cliff?) Do we need to be able to toggle it? 

I will see if I can ask @simonw if he had a cycle to look and advise. 

This is pretty cool though. A couple of defensive questions (which are no more than that): 

* Are there gotchas to this pattern we need to handle before we just ship it?
* This is entirely internal to `SimpleListFilter` — should it live in an external package first pass to iron out issues?


**Update**: I edited my comment on the ticket but, could we make this work by default (with perhaps a `?...&facets=1` toggle) for even the simple filters from a field name? 🤔

sarahboyce commented:
Hi @carltongibson thank you for testing :tada: glad it kinda works 

I think it should be possible to have the facet counts magically appear everywhere. I really like the toggle idea 
I have updated the `SimpleListFilter` one, you can have a look if you like but otherwise I will every now and again push up for the other filters. I will mark the PR as draft

I'm thinking we can have a warning that there will be a performance hit, especially for the `SimpleListFilter`, with the other ones, should be able to do some fancy aggregates to reduce the impact. :thinking: 

Thank you again :star:

sarahboyce commented:
Alright, I think that is ready for another look :+1: 
I think I could add a picture similar to at the top of the filter docs to show the filters, is there a demo site the images are coming from? 

> Are there gotchas to this pattern we need to handle before we just ship it?

If anyone has a filter with a parameter name of "facets" I think they will lose their filter, I could make the name really obscure, but it is optional

simonw commented:
> If anyone has a filter with a parameter name of "facets" I think they will lose their filter, I could make the name really obscure, but it is optional

How about using `?_facets=1` instead? That's a pattern I use a lot in Datasette - having an underscore to avoid accidental collisions with existing columns - but I don't think it's a pattern Django has used before, so it might be weird to add it for this.

simonw commented:
I'm trying this out against an existing Django application. It looks _really_ good!

I installed the branch with:

    pip install https://github.com/sarahboyce/django/archive/refs/heads/ticket_32539_lookups.zip

Now when I add `?facets=1` to an admin URL for anything that uses `list_filter=...` I get this:

<img width="936" alt="image" src="https://user-images.githubusercontent.com/9599/217049516-26469710-ce19-4bff-b3d4-527c72da7e98.png">

Django Debug Toolbar shows me that the SQL query looks like this:

<img width="444" alt="image" src="https://user-images.githubusercontent.com/9599/217049640-5c387769-b777-4b88-a0ed-4381172329b9.png">

```sql
SELECT COUNT("feedstats_subscribercount"."id") FILTER (WHERE "feedstats_subscribercount"."path" = '/atom/entries/') AS "0__c",
       COUNT("feedstats_subscribercount"."id") FILTER (WHERE "feedstats_subscribercount"."path" = '/atom/everything/') AS "1__c",
       COUNT("feedstats_subscribercount"."id") FILTER (WHERE "feedstats_subscribercount"."path" = '/atom/links/') AS "2__c"
  FROM "feedstats_subscribercount"
```
This adds a `count / filter where` clause for every displayed value. That's fine for things with only a few unique values, but it does mean you end up with a pretty huge SQL query for columns with a large number of values. Arguably those aren't a great fit for the `list_filter=` function though, so I'm not sure if that's something we need to worry about.

I like the idea of a toggle to turn these on and off. I would want them on by default for most of my projects - I wonder if the Django `admin.ModelAdmin` subclasses could grow a `show_facets=True` option that sets that default?

simonw commented:
I noticed that it's not currently sorting the listed filters by the number of results:

<img width="219" alt="image" src="https://user-images.githubusercontent.com/9599/217051688-ddfbef0f-e566-45ae-8cd8-e724d994179f.png">

When I'm working with facet interfaces generally I always like to sort them with the largest counts at the top, as then the feature doubles as a "what's most popular" feature.

This makes more sense for things like column value facets than it does for date filters - this filter interface here makes more sense sticking to its time order than sorting by those numbers:

<img width="180" alt="image" src="https://user-images.githubusercontent.com/9599/217051877-859f47ec-fb0b-4d32-8b12-24f77556c532.png">

It looks to me like Django defaults to sorting these filters alphabetically - as illustrated by this larger example that facets/filters by user agent:

<img width="269" alt="image" src="https://user-images.githubusercontent.com/9599/217052105-b2f74d61-53b0-413e-a643-f7fdbf124819.png">

I don't have a particularly strong opinion about this. I feel that sorting by numerical count makes more sense that sorting alphabetically for most cases, but that's just an intuition based on how I've used these kinds of interfaces in the past - I'm not ready to confidently state that every Django project in existence would benefit more from numerical than from alphabetical sorting here.

simonw commented:
Just spotted this: clicking on one of those filter links currently does _not_ persist the `?facets=1` in the URL:

![facets](https://user-images.githubusercontent.com/9599/217052976-b4b45af5-6718-422b-9517-b15c436c8c94.gif)

This feels wrong to me: when I'm interacting with facets I like to do things like drill down using one to see how that affects the numeric counts for the others.

Ideally the URLs generated for the facets should persist the `?facets=1` query string parameter.

simonw commented:
I'm really impressed with how slim this implementation is - love how much new functionality we get from so little code, and the tests look good to me too.

kezabelle commented:
> How about using `?_facets=1` instead? [...] having an underscore to avoid accidental collisions with existing columns - but I don't think it's a pattern Django has used before [...]

👍 There is some prior precedent -- `_popup` and `_to_field` keys are already used in `contrib.admin.options`.

sarahboyce commented:
Thank you @simonw for the review and testing :star: 
- `facets` has been updated to `_facets`
- adding `_facets=1` to the url should now be persisted in filter links
- added an attribute `show_facets` to the ModelAdmin to allow facets to be turned on by default without using the query parameter
- I did an update for getting the counts for `SimpleListFilter` to achieve this in a single query
- Updated the docs and tried to add a note around that the queries may cause performance issues

I need to think a bit more about the ordering, I can see the value of having it. Maybe the `SimpleListFilter` as it's set up with a specific order _maybe_ doesn't want to be ordered (but maybe does) and the date one mostly will look ok with the count order just sometimes 7 days ago is bigger than this year. Choosing to add it to some and not others feels like it might look like a bug. So I feel like it should be all or nothing.


I think this is ready for another set of testing and review, happy for anyone else to weigh in their thoughts on whether they are for or against the ordering also.

pauloxnet commented:
@sarahboyce congratulations for the functionality you propose and for its implementation. I can only suggest to use more f-strings in some places of the code because it's more readable and faster. Also, I think the Django version should be specified in the two additions to the documentation.

smithdc1 commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -210,3 +210,14 @@ It is possible to specify a custom template for rendering a list filter::
+
+Counts can be appended in brackets to each of the filter option labels by
+adding ``_facets=1`` to the URL query. These counts will update according to the
+current applied filters. To have these applied by default for a model admin,
```
```suggestion
currently applied filters. To have these applied by default for a model admin,
```

smithdc1 commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -965,6 +965,22 @@ subclass::
+    .. versionadded:: 5.0
+
+    Set ``show_facets`` to ``True`` to have all filter options appended with
+    the count of results that match that filter option in brackets. These
```
```suggestion
    the count of results that match that filter option. These
```

It took me a few attempts to understand the 'in brackets' part. Maybe we can drop that part? 🤷‍♂️

sarahboyce replied:
Yeah no problem :+1:

smithdc1 commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -965,6 +965,22 @@ subclass::
+    counts will update in line with currently applied filters. By default, this
+    is set to ``False``.
+
+    .. admonition:: Performance considerations with facets
```
```suggestion
    .. admonition:: Performance considerations with facets.
```
Does this need a full stop?

sarahboyce replied:
A lot of the other admonitions don't have full stops so I guess maybe it's like a title and not needed? I'm not sure though, happy to add it in

felixxm commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -210,3 +210,14 @@ It is possible to specify a custom template for rendering a list filter::
 
 See the default template provided by Django (``admin/filter.html``) for a
 concrete example.
+
```
Chop unnecessary blank line.
```suggestion
```

felixxm commented on tests/admin_filters/tests.py:
```diff
@@ -1315,6 +1333,307 @@ def test_simplelistfilter_with_queryset_based_lookups(self):
+        request.user = self.alfred
+        changelist = modeladmin.get_changelist_instance(request)
+        queryset = changelist.get_queryset(request)
+        self.assertEqual(list(queryset), list(Book.objects.order_by("-id")))
```
```suggestion
        self.assertSequenceEqual(queryset, Book.objects.order_by("-id"))
```

carltongibson reviewed (commented):
This is looking great. 

> I need to think a bit more about the ordering...

Unless you have a solution come to mind @sarahboyce (which you might! 😅) I'd be inclined towards taking ordering as a possible follow-up ticket, rather than blocking on it. (Most of the filters I use don't have too many values, since I generally switch to search once it gets out of hand, and this is already a massive win — headline feature for the release likely.)

carltongibson commented on django/contrib/admin/views/main.py:
```diff
@@ -103,6 +111,7 @@ def __init__(
             self.page_num = 1
         self.show_all = ALL_VAR in request.GET
         self.is_popup = IS_POPUP_VAR in request.GET
+        self.add_facets = model_admin.show_facets or IS_FACETS_VAR in request.GET
```
I just wonder if this should be an enum `{ALWAYS|ALLOW|NEVER}` (or something) — with `NEVER` disabling the faceting entirely. 

(_"One of our admin users brought down production by adding the `IS_FACETS_VAR` query string arg..."_)


🤔

sarahboyce replied:
Ooooh this is a clever idea, with `NEVER` as the default?

carltongibson replied:
I thought `ALLOW` as the default.

sarahboyce replied:
That's updated, though might not be documented in the clearest of ways. If there is a nice setting based off an enum I can look at, let me know. Happy to update :)

sarahboyce commented:
> Unless you have a solution come to mind @sarahboyce (which you might! sweat_smile) 

I have _something_ for the ordering [9c0f1f5](https://github.com/django/django/pull/16495/commits/9c0f1f5ba4361fad14c60c62ea11a9a7772176d9), I have put it in a seperate commit and we can still decide whether to include it or not. :+1:

sarahboyce commented on docs/ref/contrib/admin/filters.txt:
```diff
@@ -210,3 +210,14 @@ It is possible to specify a custom template for rendering a list filter::
+.. versionadded:: 5.0
+
+Counts can be appended in brackets to each of the filter option labels by
+adding ``_facets=1`` to the URL query. These counts will update according to
```
You're right :+1: updated

sarahboyce commented on django/contrib/admin/options.py:
```diff
@@ -69,6 +69,13 @@
+class FacetFilterSetting(models.TextField):
+    NEVER = "NEVER"
+    ALLOW = "ALLOW"
+    ALWAYS = "ALWAYS"
```
> [0]: I think. Maybe folks would want it separately documented to search for/link to. (Not sure.)

I think I know what you're after but not sure the most natural place to put it/how best to document it :thinking: 
I have made the other updates

carltongibson replied:
Ok, no stress. Let me take a pop tomorrow AM with a concrete suggestion. 👍

carltongibson reviewed (commented):
Hi @sarahboyce. 

OK I pushed some suggested docs edits. Please take a look, and use as you think. (The main bit was documenting the `ShowFacets` enum as nested in the `show_facets` section, since that's the only place it applies — but it's targetable with rst roles this way.) 

I'm not sure if `change-list` is meant to have a hyphen or not. Sorry 😅


Then I just wonder — please feel free to shout "NO!" at me :) — if a UI toggle to add/remove the `_facets` query parameter just at the top of the filter list might be a nice finishing touch? 🤔 (For me, it's a **long way** until 5.0, and we could add that as an easy picking ticket for someone to follow-up with...)

carltongibson commented on docs/releases/5.0.txt:
```diff
@@ -46,6 +46,10 @@ Minor features
 * The new :meth:`.AdminSite.get_log_entries` method allows customizing the
   queryset for the site's listed log entries.
 
+* Facet counts are now shown for applied filters in admin change-list UI. By
+  default, facet counts can be toggled via a ``_facets`` query parameter, but
+  this behavior can be configured via the new :attr:`.ModelAdmin.show_facets`.
+
```
This is a headline feature for me.

felixxm replied:
Agreed.

carltongibson commented on django/contrib/admin/views/main.py:
```diff
@@ -103,6 +112,9 @@ def __init__(
             self.page_num = 1
         self.show_all = ALL_VAR in request.GET
         self.is_popup = IS_POPUP_VAR in request.GET
+        self.add_facets = model_admin.show_facets == ShowFacets.ALWAYS or (
+            model_admin.show_facets == ShowFacets.ALLOW and IS_FACETS_VAR in request.GET
```
Enum comparisons should use `is`: https://docs.python.org/3/howto/enum.html#comparisons

felixxm commented on django/contrib/admin/filters.py:
```diff
@@ -312,25 +399,48 @@ def __init__(self, field, request, params, model, model_admin, field_path):
-        for lookup, title in self.field.flatchoices:
+        none_count = None
+        for i, (lookup, title) in enumerate(self.field.flatchoices):
+            count = facet_counts[f"{i}__c"] if add_facets else None
+            title = f"{title} ({count})" if add_facets else title
```
```suggestion
            if add_facets:
                count = facet_counts[f"{i}__c"]
                title = f"{title} ({count})"
            else:
                count = None
```

sarahboyce replied:
Tried to reduce the ternaries :+1:

sarahboyce commented:
> OK I pushed some suggested docs edits. Please take a look, and use as you think. (The main bit was documenting the `ShowFacets` enum as nested in the `show_facets` section, since that's the only place it applies — but it's targetable with rst roles this way.)

It looks good to me! Thank you, the enum especially is much clearer :star: 

> Then I just wonder — please feel free to shout "NO!" at me :) — if a UI toggle to add/remove the `_facets` query parameter just at the top of the filter list might be a nice finishing touch? thinking (For me, it's a **long way** until 5.0, and we could add that as an easy picking ticket for someone to follow-up with...)

Ahhhh I see what you mean, it's definitely nice for feature discovery :thinking: probably quite important for it to be used
Similar to the pretty toggle for the admin light mode/dark mode. I will have a quick look later and if I'm struggling, I'll push to another ticket :+1:

felixxm commented on django/contrib/admin/views/main.py:
```diff
@@ -114,7 +126,6 @@ def __init__(
             del self.params[PAGE_VAR]
         if ERROR_FLAG in self.params:
             del self.params[ERROR_FLAG]
-
```
Please revert unrelated formatting change.

felixxm commented on django/contrib/admin/filters.py:
```diff
@@ -454,34 +581,59 @@ def __init__(self, field, request, params, model, model_admin, field_path):
+        empty_title = self.empty_value_display
+        empty_count = None
+        for i, val in enumerate(self.lookup_choices):
+            count = facet_counts[f"{i}__c"] if add_facets else None
```
I would format this as in `ChoicesFieldListFilter` to avoid so many ternaries.

felixxm commented on django/contrib/admin/filters.py:
```diff
@@ -376,37 +487,37 @@ def __init__(self, field, request, params, model, model_admin, field_path):
-                    self.lookup_kwarg_since: str(today),
-                    self.lookup_kwarg_until: str(tomorrow),
+                    self.lookup_kwarg_since: today,
+                    self.lookup_kwarg_until: tomorrow,
```
Moving `str()` calls to the `DateFieldListFilter.get_choices()` sound like a cleanup that can be done in advance :thinking: What do you think?

sarahboyce replied:
[fbed3f1](https://github.com/django/django/pull/16495/commits/fbed3f1283ba45259a6fe252defc8e9cc5b55d66) pulled it out into it's own commit :+1:

felixxm reviewed (commented):
@sarahboyce Many thanks :star: I left small suggestions.

felixxm commented on django/contrib/admin/filters.py:
```diff
@@ -111,19 +126,45 @@ def lookups(self, request, model_admin):
-        for lookup, title in self.lookup_choices:
+        for i, (lookup, title) in enumerate(self.lookup_choices):
+            if add_facets:
+                count = facet_counts.get(f"{i}__c", -1)
+                if count != -1:
+                    title = f"{title} ({count})"
+                else:
+                    title = f"{title} (-)"
```
Maybe walrus :smiley: Sorry for nitpicking
```suggestion
                if (count := facet_counts.get(f"{i}__c", -1)) != -1:
                    title = f"{title} ({count})"
                else:
                    title = f"{title} (-)"
```

sarahboyce replied:
No worries :+1: updated

sarahboyce commented:
Toggle off
![image](https://user-images.githubusercontent.com/42296566/219455041-5eba03bc-358c-4438-9b97-9f5f9e83db7f.png)

Toggle on
![image](https://user-images.githubusercontent.com/42296566/219455179-7764037a-62ce-40c1-a0c5-e9c4d27cda5b.png)

:shrug: maybe?

akshaykedar1996 commented:
I'd be happy to provide more information on the change made in Django's codebase!

In Django, the ModelAdmin class is used to define the admin interface for a particular model. Prior to the change made in ticket #16495, ModelAdmin did not have built-in support for facet filters, which allow users to filter data based on a set of predefined options.

In ticket #32539, support for toggleable facet filters was added to ModelAdmin. This change allows developers to specify a list of facets for a particular model, along with the options for each facet. Users can then toggle the facets to filter the data based on their preferences.

To add facet filters to a ModelAdmin class, developers can use the list_filter attribute and specify the facets as a list of strings. For example, to add facet filters for a Book model based on the author and genre fields, the following code can be used:

python

Models.py file
-----------------------------------------------------------------------------------------------------------------------
class BookAdmin(admin.ModelAdmin):
    list_filter = ('author', 'genre')
------------------------------------------------------------------------------------------------------------------------
By default, the facet filter options are displayed in a list, but developers can customize the appearance and behavior of the facet filters using the list_filter_template attribute.

Overall, the addition of toggleable facet filters to ModelAdmin provides a more user-friendly interface for filtering data in the Django admin, which is an important feature for many Django applications.

felixxm commented on tests/admin_filters/tests.py:
```diff
@@ -1150,7 +1172,7 @@ def test_simplelistfilter(self):
         # Make sure the correct queryset is returned
         queryset = changelist.get_queryset(request)
-        self.assertEqual(list(queryset), list(Book.objects.order_by("-id")))
+        self.assertSequenceEqual(list(queryset), list(Book.objects.order_by("-id")))
```
`list()` is unnecessary when you use `assertSequenceEqual()`:
```suggestion
        self.assertSequenceEqual(queryset, Book.objects.order_by("-id"))
```

sarahboyce replied:
`self.assertSequenceEqual(queryset, list(Book.objects.order_by("-id")))`
I can remove the `list()` from queryset, but I need it wrapping the other one. Not sure why though, to have the query evaluated maybe? :thinking:

felixxm replied:
Ah yes because we have two querysets :+1:

felixxm commented on tests/admin_filters/tests.py:
```diff
@@ -1315,6 +1337,390 @@ def test_simplelistfilter_with_queryset_based_lookups(self):
+    def test_filters_with_facets(self):
+        modeladmin = DecadeFilterBookAdmin(Book, site)
+
+        # Facets with no filters applied
```
```suggestion
        # Facets with no filters applied.
```

felixxm commented on tests/admin_filters/tests.py:
```diff
@@ -1315,6 +1337,390 @@ def test_simplelistfilter_with_queryset_based_lookups(self):
+        request.user = self.alfred
+        changelist = modeladmin.get_changelist_instance(request)
+        queryset = changelist.get_queryset(request)
+        self.assertSequenceEqual(list(queryset), list(Book.objects.order_by("-id")))
```
```suggestion
        self.assertSequenceEqual(queryset, Book.objects.order_by("-id"))
```

felixxm commented on tests/admin_filters/tests.py:
```diff
@@ -1315,6 +1337,390 @@ def test_simplelistfilter_with_queryset_based_lookups(self):
+        request.user = self.alfred
+        changelist = modeladmin.get_changelist_instance(request)
+        queryset = changelist.get_queryset(request)
+        self.assertEqual(list(queryset), [])
```
```suggestion
        self.assertSequenceEqual(queryset, [])
```

felixxm commented on tests/admin_filters/tests.py:
```diff
@@ -1315,6 +1337,390 @@ def test_simplelistfilter_with_queryset_based_lookups(self):
+        request.user = self.alfred
+        changelist = modeladmin.get_changelist_instance(request)
+        queryset = changelist.get_queryset(request)
+        self.assertEqual(
+            list(queryset),
+            list(Book.objects.filter(author=self.alfred).order_by("-id")),
+        )
```
```suggestion
        self.assertSequenceEqual(
            queryset, Book.objects.filter(author=self.alfred).order_by("-id")
        )
```

felixxm commented on tests/admin_filters/tests.py:
```diff
@@ -1315,6 +1337,390 @@ def test_simplelistfilter_with_queryset_based_lookups(self):
+        request.user = self.alfred
+        changelist = modeladmin.get_changelist_instance(request)
+        queryset = changelist.get_queryset(request)
+        self.assertSequenceEqual(list(queryset), list(Book.objects.order_by("-id")))
```
```suggestion
        self.assertSequenceEqual(queryset, Book.objects.order_by("-id"))
```

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -807,6 +807,36 @@ def test_limited_filter(self):
         )
 
+    def test_change_list_facet_toggle(self):
+        # Toggle visible when show_facet is the default of admin.ShowFacets.ALLOW
```
Wrap comments at 79 chars
```suggestion
        # Toggle is visible when show_facet is the default of
        # admin.ShowFacets.ALLOW.
```

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -807,6 +807,36 @@ def test_limited_filter(self):
+            '<a href="?" class="hidelink">Remove Facets</a>',
+            msg_prefix="Expected facet filter toggle not found in changelist view",
+        )
+        # Toggle not visible when show_facet is admin.ShowFacets.ALWAYS
```
```suggestion
        # Toggle is not visible when show_facet is admin.ShowFacets.ALWAYS.
```

felixxm commented on tests/admin_views/tests.py:
```diff
@@ -807,6 +807,36 @@ def test_limited_filter(self):
+            "Facets",
+            msg_prefix="Expected not to find facet toggle in changelist view",
+        )
+        # Toggle not visible when show_facet is admin.ShowFacets.NEVER
```
```suggestion
        # Toggle is not visible when show_facet is admin.ShowFacets.NEVER.
```

felixxm commented on tests/admin_filters/tests.py:
```diff
@@ -1315,6 +1337,390 @@ def test_simplelistfilter_with_queryset_based_lookups(self):
+
+        filterspec_simplelistfilter = changelist.get_filters(request)[0][1]
+        choices = list(filterspec_simplelistfilter.choices(changelist))
+        self.assertEqual(choices[1]["display"], "the 2000's (2)")
+        self.assertEqual(choices[2]["display"], "the 1990's (1)")
+        self.assertEqual(choices[3]["display"], "the 1980's (0)")
+        self.assertEqual(choices[4]["display"], "other decades (-)")
```
Can we add a hook for checking `choices`? For example:
```python
def assertChoicesDisplay(self, choices, expected_displays):
    for choice, expected_display in zip(choices[1:], expected_displays, strict=True):
        self.assertEqual(choice["display"], expected_display])
```
and use it as:
```python
    self.assertChoicesDisplay(
        choices, ["the 2000's (2)", "the 1990's (1)", "the 1980's (0)", "other decades (-)"]
    )
```

sarahboyce replied:
Nice idea :+1:

felixxm commented on docs/releases/5.0.txt:
```diff
@@ -37,6 +37,13 @@ compatible with Django 5.0.
+Facet filters in the admin
+--------------------------
+
+Facet counts are now shown for applied filters in admin change-list UI. By
```
```suggestion
Facet counts are now shown for applied filters in the admin changelist. By
```

felixxm commented on docs/releases/5.0.txt:
```diff
@@ -37,6 +37,13 @@ compatible with Django 5.0.
+
+Facet counts are now shown for applied filters in admin change-list UI. By
+default, facet counts can be toggled via a ``_facets`` query parameter, but
+this behavior can be configured via the new :attr:`.ModelAdmin.show_facets`.
```
```suggestion
this behavior can be changed via the new :attr:`.ModelAdmin.show_facets` attribute.
```

felixxm commented on docs/releases/5.0.txt:
```diff
@@ -37,6 +37,13 @@ compatible with Django 5.0.
+Facet counts are now shown for applied filters in admin change-list UI. By
+default, facet counts can be toggled via a ``_facets`` query parameter, but
+this behavior can be configured via the new :attr:`.ModelAdmin.show_facets`.
+
```
I'd add a reference to the `Facets` section in the `docs/ref/contrib/admin/filters.txt`

felixxm commented on docs/ref/contrib/admin/index.txt:
```diff
@@ -965,6 +965,54 @@ subclass::
+
+    .. versionadded:: 5.0
+
+    Controls whether facet counts are displayed for filters in the change-list
+    UI. Default: :attr:`.ShowFacets.ALLOW`.
```
```suggestion
    Controls whether facet counts are displayed for filters in the admin changelist
    Defaults to :attr:`.ShowFacets.ALLOW`.
```

felixxm reviewed (commented):
@sarahboyce Thanks :+1: I left small suggestions.

Is `_facets` preserved as changelist filters?

carltongibson commented on docs/releases/5.0.txt:
```diff
@@ -37,6 +37,13 @@ compatible with Django 5.0.
+--------------------------
+
+Facet counts are now shown for applied filters in admin change-list UI. By
+default, facet counts can be toggled via a ``_facets`` query parameter, but
```
Just one query: do we want to say, "can be toggled in the UI" now, rather than mentioning the query parameter? (Here and in docs.)

sarahboyce replied:
Yes true, it probably is less confusing also - I've tried to update


There is a picture at the top here which I would like to update, I don't suppose this is someones test app?
![image](https://user-images.githubusercontent.com/42296566/220895184-dc7dec6f-37e4-468b-ae5e-785fc94d97ab.png)

https://docs.djangoproject.com/en/4.1/ref/contrib/admin/filters/


I saw there is this ticket: https://code.djangoproject.com/ticket/34043 which would be generally useful and I could also use for updating documentation screenshots. Might be a good one for a Django sprint maybe

simonw replied:
I have a pretty elaborate mechanism for automated screenshots for my own project, I wonder if Django might benefit from something similar?

Definitely not as a blocker for this change though!

https://simonwillison.net/2022/Oct/14/automating-screenshots/

sarahboyce commented:
> Is `_facets` preserved as changelist filters?

It's preserved when clicking on other filter options (so the issue described earlier). I think it wouldn't preserve if you do things like click on another table/changelist view or click on the breadcrumbs.

alewish2196 commented:
Your email has been received and will be addressed by our team shortly.

On Thu, Feb 23, 2023 at 7:02 AM Sarah Boyce ***@***.***>
wrote:

> Is _facets preserved as changelist filters?
>
> It's preserved when clicking on other filter options (so the issue
> described earlier). I think it wouldn't preserve if you do things like
> click on another table/changelist view or click on the breadcrumbs.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/django/django/pull/16495#issuecomment-1441641598>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG6RARSXPHKQ4YTKBJ7TEJDWY5GUFANCNFSM6AAAAAAUCRCIXA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

alewish2196 commented:
I received your email and wanted to confirm that it has been received and
is being reviewed.

On Thu, Feb 23, 2023 at 6:55 AM Mariusz Felisiak ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In tests/admin_filters/tests.py
> <https://github.com/django/django/pull/16495#discussion_r1115567552>:
>
> > @@ -1150,7 +1172,7 @@ def test_simplelistfilter(self):
>
>
>
>          # Make sure the correct queryset is returned
>
>          queryset = changelist.get_queryset(request)
>
> -        self.assertEqual(list(queryset), list(Book.objects.order_by("-id")))
>
> +        self.assertSequenceEqual(list(queryset), list(Book.objects.order_by("-id")))
>
>
> Ah yes because we have two querysets 👍
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/django/django/pull/16495#discussion_r1115567552>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AG6RARV2MULO7MOUHXAYJS3WY5FZJANCNFSM6AAAAAAUCRCIXA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

simonw commented:
> Toggle off ![image](https://user-images.githubusercontent.com/42296566/219455041-5eba03bc-358c-4438-9b97-9f5f9e83db7f.png)
> 
> Toggle on ![image](https://user-images.githubusercontent.com/42296566/219455179-7764037a-62ce-40c1-a0c5-e9c4d27cda5b.png)
> 
> 🤷 maybe?

I think "show counts / hide counts" would be clearer for the button text here.

pauloxnet commented:
> I think "show counts / hide counts" would be clearer for the button text here.

I agree.

sarahboyce commented:
Thank you :+1: updated

Should I also localize the numbers? So `Test (1000)` looks like `Test (1,000)` / `Test (1.000)`
I think I need to use something like `formats.localize(count, use_l10n=True)` everywhere if we want it

felixxm commented:
> > Is `_facets` preserved as changelist filters?
> 
> It's preserved when clicking on other filter options (so the issue described earlier). I think it wouldn't preserve if you do things like click on another table/changelist view or click on the breadcrumbs.

It looks like it it preserved as a part of changelist filters (`_changelist_filters`) :+1:

felixxm commented:
`Hide/Show counts` should have the same font size as `Clear all filters`
![image](https://user-images.githubusercontent.com/2865885/221113872-d888ee91-2216-4d0a-b849-42b6792c2f92.png)
We should also wrap both in `div` and move the bottom border from `#changelist-filter #changelist-filter-clear a` to the new `div`.

felixxm commented:
> > Unless you have a solution come to mind @sarahboyce (which you might! sweat_smile)
> 
> I have _something_ for the ordering [9c0f1f5](https://github.com/django/django/pull/16495/commits/9c0f1f5ba4361fad14c60c62ea11a9a7772176d9), I have put it in a seperate commit and we can still decide whether to include it or not. +1

I found ordering by facets quite disturbing as the list of choices is constantly changing. I'm also not sure what to do with facets when an option is selected because then all other options appear with `(0)`. It's confusing that changing the filter to an option with `(0)` shows a non-empty list. We should probably ignore the selected filter when calculating facets for the same option :thinking:

https://user-images.githubusercontent.com/2865885/221117042-45fe0c1f-daa6-4507-903a-e17993da994d.mp4

felixxm commented on django/contrib/admin/templates/admin/change_list.html:
```diff
@@ -75,9 +75,14 @@
-            </h3>{% endif %}
+            {% if cl.is_facets_optional or cl.has_active_filters %}<div id="changelist-filter-extra-actions">
+              {% if cl.is_facets_optional %}<h3 id="changelist-filter-facets">
+                {% if cl.add_facets %}<a href="{{ cl.remove_facet_link }}" class="hidelink">{% translate 'Hide counts' %}</a>{% else %}<a href="{{ cl.add_facet_link }}">{% translate 'Show counts' %}</a>{% endif %}
```
`viewlink` disappeared:
```suggestion
                {% if cl.add_facets %}<a href="{{ cl.remove_facet_link }}" class="hidelink">{% translate 'Hide counts' %}</a>{% else %}<a href="{{ cl.add_facet_link }}" class="viewlink">{% translate 'Show counts' %}</a>{% endif %}
```

sarahboyce commented:
![image](https://user-images.githubusercontent.com/42296566/221127283-0a665d97-a22d-4495-ad3c-74800933e949.png)
The size is updated :+1: nice suggestion

Also made it not include the current filter option when one of same group applied (screenshot hopefully illustrates it) but would appreciate feedback on that bit as feels a little hacky.

 On the reordering, I can remove it :shrug: I don't really mind, I added it based of @simonw 's suggestion here but also
> I don't have a particularly strong opinion about this. I feel that sorting by numerical count makes more sense that sorting alphabetically for most cases, but that's just an intuition based on how I've used these kinds of interfaces in the past - I'm not ready to confidently state that every Django project in existence would benefit more from numerical than from alphabetical sorting here.

I guess with the adjustment that it doesn't apply to the one you selected, it should move a little less (only the ones you weren't looking at move) maybe it's now less disturbing? Or is it still jarring?

felixxm commented:
@sarahboyce Can you drop reordering (at least for now)? This should simplify the current implementation, e.g. `OptionalFacetsMixin` would be unnecessary.

I'd like to merge it :star2:

sarahboyce commented:
It should be dropped :+1:

felixxm commented:
@sarahboyce I merged the first commit in 4cb55733527d39ebf58c1987f72a19b7c2e0bac6 and pushed small edits to a few files.

I noticed an issue while polishing this patch, i.e. `facets` is not preserved when using `X total` option:


https://user-images.githubusercontent.com/2865885/222430022-c0bffc24-31ff-4ebf-a08c-66590ab7ddf8.mp4



Can you take a look? Please push changes in a separate commit.

sarahboyce commented:
Well spotted :+1: thank you for testing :star:

felixxm commented:
@sarahboyce I tried to move some common logic to the `FacetsMixin`, what do you think?

felixxm commented on django/contrib/admin/filters.py:
```diff
@@ -269,7 +269,7 @@ def get_facet_counts(self, pk_attname, filtered_qs):
         }
         if self.include_empty_choice:
-            counts["empty__c"] = models.Count(
+            counts["__c"] = models.Count(
```
Using `empty__c` could potentially conflict with the _"empty"_ primary key.

sarahboyce commented:
> @sarahboyce I tried to move some common logic to the `FacetsMixin`, what do you think?

I like it :+1:

smithdc1 commented on docs/releases/5.0.txt:
```diff
@@ -37,6 +37,14 @@ compatible with Django 5.0.
+
+Facet counts are now shown for applied filters in the admin changelist when
+toggled on via the UI. This behavior can be changed via the new
+:attr:`.ModelAdmin.show_facets` attribute. For more information see
```
Maybe the toggle in the UI should be referred to as headline feature rather than the `ModelAdmin` attribute?

felixxm replied:
@smithdc1 Any wording suggestion?

smithdc1 replied:
I must have completely miss-read this earlier today 🤦.

The first sentence references the UI so I think it's fine as it is (for some reason I thought this section made no reference to the UI toggle at all). 

Sorry.

felixxm commented:
I reorganized and simplified tests with `subTest()`, we're getting close... :crossed_fingers:

felixxm commented:
@sarahboyce Thanks for all your efforts :1st_place_medal: I updated screenshots and this is ready to merge! :tada:

felixxm commented:
buildbot, test on selenium.

simonw commented:
I deployed this on my blog and it's working great!

<img width="1131" alt="image" src="https://user-images.githubusercontent.com/9599/223229422-a59d8ced-3ea1-4a0a-b074-8ad9ca510618.png">
